packit_service/models.py

```diff
@@ -133,6 +134,80 @@ def optional_timestamp(datetime_object: Optional[datetime]) -> Optional[int]:
     return None if datetime_object is None else int(datetime_object.timestamp())
 
 
+def get_submitted_time_from_model(
+    model: Union["CoprBuildTargetModel", "TFTTestRunTargetModel"]
+) -> datetime:
+    # TODO: unify `submitted_name` (or better -> create for both models `task_accepted_time`)
+    # to delete this mess plz
+    if isinstance(model, CoprBuildTargetModel):
+        return model.build_submitted_time
+
+    return model.submitted_time
+
+
+def get_most_recent_targets(
+    models: Union[
+        Iterable["CoprBuildTargetModel"],
+        Iterable["TFTTestRunTargetModel"],
+    ],
+) -> List[Union["CoprBuildTargetModel", "TFTTestRunTargetModel"]]:
+    """
+    Gets most recent models from an iterable (regarding submission time).
+
+    Args:
+        models: Copr or TF models - if there are any duplicates in them then use the most
+         recent model
+
+    Returns:
+        Dictionary - target as a key and corresponding most recent model as a value.
+    """
+    most_recent_models: Dict[
+        str, Union["CoprBuildTargetModel", "TFTTestRunTargetModel"]
+    ] = {}
+    for model in models:
+        submitted_time_of_current_model = get_submitted_time_from_model(model)
+        if (
+            most_recent_models.get(model.target) is None
+            or get_submitted_time_from_model(most_recent_models[model.target])
+            < submitted_time_of_current_model
+        ):
+            most_recent_models[model.target] = model
+
+    return list(most_recent_models.values())
+
+
+def filter_most_recent_target_models_by_status(
+    models: Union[
+        Iterable["CoprBuildTargetModel"],
+        Iterable["TFTTestRunTargetModel"],
+    ],
+    statuses_to_filter_with: List[str],
+) -> Optional[Set[Union["CoprBuildTargetModel", "TFTTestRunTargetModel"]]]:
```

Comment:
Given the input, I guess you meant

```suggestion
) -> Optional[Union[Set["CoprBuildTargetModel"], Set["TFTTestRunTargetModel"]]]:
```

Reply:
I haven't written this code, but yes, the types were wrong and not only here...I tried to correct it (1. commit), but I don't know how to explain some things to mypy (line 226 in models.py :D)..I would be glad for any advice